WebUI: Introduce 'By Shown File Order' priority in WebUI

userwiths opened the pull request:
Closes #16845 

## Issue summary

There is an open in the context menu of the files panel, that allows assigning priority to the selected files `By Shown File Order`. This option exists in the native/desktop application but is missing from the WebUI interface.

## Summary of changes

Added new priority option and refactored the event that is triggered upon selecting an option.

The "refactored" event is entirely JS code.
The approach I used is supposed to handle both the general case (a single priority for a group of files) and the `By Shown File Order` case (where we determine the priority based on the number of files and their positioning).

We do this by having an object with the value of the `priority` as a key and as value a list of `row_id`/`file_id` pairs. We keep both the `row` & `file` ids because `setFilePriority` wants both of them.

If the change is on one of the "general" priorities (Do Not Download, Normal, High, Maximum) all files/rows go into the same group, otherwise they are distributed by the same logic as in [the desktop application](https://github.com/qbittorrent/qBittorrent/blob/master/src/gui/torrentcontentwidget.cpp#L282)

## How did you test ?

Manually with a torrent containing exactly 10 files. Dont have a larger one at hand atm.

userwiths commented on src/webui/www/private/scripts/prop-files.js:
```diff
@@ -478,7 +478,54 @@ window.qBittorrent.PropFiles ??= (() => {
         }
 
-        setFilePriority(Object.keys(uniqueRowIds), Object.keys(uniqueFileIds), priority);
+        // Will be used both in general case and ByShownOrder case
```
I think most of the `deletions` in the PR are coming from changes ive done regarding this comment.

Im gonna leave my notes bellow in case they could help to understand my changes:

---


1. Made a recursive function with the clear and only intent of collecting `fileIds` of the currently selected files and all their children in case its a directory.
2. Removed the need for `rowIds` (almost) entirely. The almost is there because folders have a `fileId` of `-1` so I use the `rowId` in order to enumerate their children.
    * Removal of `rowIds` was done by changing `setFilePriority` where the `ignore` logic (which calls `setIgnored` from `TorrentFilesTable`) was the only part that required them.
    * The `remaining` value (which was updated by `setIgnored`) should already be covered in `handleNewTorrentFiles` function called by `loadTorrentFilesData` after priority change.

Chocobo1 replied:
Excuse me, I don't really see the point. In addition, I think your new code made it more complex than it need to be.

userwiths replied:
> I would suggest split the code path. Current code is too cumbersome IMO.

Maybe i misunderstood this request, mind clarifying what you advice me to do here ?

> Excuse me, I don't really see the point

When you mentioned "cumbersome code" the first thing that came to mind for me were the variables `uniqueFileIds` & `uniqueRowIds` (unique ? so we can have duplicates even tho we cant select a file twice ?) and the fact that we use them as objects/dictionaries only to pass them with `Object.keys` (values are never used?).
So in my head having **only** `fileIds` and putting the current selected + their children's `fileId`'s there was the way to go.

I'm gonna check the other comments soon, but would like some clarification on this one.

userwiths commented:
So, it seems the logic was changed a bit in master while I was away, in a positive way imo.

I rebased and adapted the logic where i believe it should be, gave it a round of testing and updated the issue with the requested screenshot.

Hopefully it is now "simple" enough, in case there are problems with the suggested changes leave me a review and im gonna get to them as soon as i can.
